engine/src/resource/json_utils.cpp

```diff
@@ -22,67 +119,136 @@
  * along with Vega Strike. If not, see <https://www.gnu.org/licenses/>.
  */
 
-#include "json_utils.h"
-
 #include <string>
 #include <vector>
+#include <stdexcept>
+
 #include <boost/algorithm/string.hpp>
 #include <boost/json.hpp>
 
+/**
+ * @brief Retrieve a nested JSON value, given a pipe-delimited path.
+ *        If no key is found at any step, returns an empty JSON value.
+ * @param object   The root JSON object.
+ * @param key      A string containing one or more subkeys, delimited by '|'.
+ * @return boost::json::value   The found value, or an empty value on failure.
+ */
+static boost::json::value JsonGetValue(const boost::json::object& object,
+                                       const std::string& key)
+{
+    std::vector<std::string> key_sections;
+    boost::split(key_sections, key, boost::is_any_of("|"));
 
+    const boost::json::object* current_obj = &object;
+    for (size_t i = 0; i < key_sections.size(); ++i) {
+        const auto& k = key_sections[i];
 
-
-
-const std::string JsonGetStringWithDefault(boost::json::object object, 
-                                           const std::string& key, 
-                                           const char* default_value) {
-    boost::json::value value = JsonGetValue(object, key);
-    if(value != nullptr) {
-        return boost::json::value_to<std::string>(value);
-    } else {
-        return default_value;
+        if (auto it = current_obj->find(k); it != current_obj->end()) {
+            // If this is the final part of the path, return the value
+            if (i == key_sections.size() - 1) {
+                return it->value();
+            }
+            // Otherwise, we expect this to be another nested object
+            if (it->value().is_object()) {
+                current_obj = &it->value().get_object();
+            } else {
+                // Path ended with a non-object before the last subkey
+                return {};
+            }
+        } else {
+            // Key not found
+            return {};
+        }
     }
+    // If we somehow get here, return empty
+    return {};
```

Comment:
This is interesting. 

`return {}`

+

`const boost::json::value val = JsonGetValue(object, key);
if (!val.is_null()) {`

will probably work. I wasn't a fan of my use of nullptr there, though this is harder to read in some ways.